cpu/gd32v/periph/i2c.c

```diff
@@ -189,227 +208,522 @@ void i2c_release(i2c_t dev)
+        res = _i2c_write_cmd(dev, buffer[i]);
+        if (res != 0) {
+            return res;
         }
     }
-    /*
-    * If timeout occurs this means a problem that must be handled on a higher
-    * level.  A SWRST is recommended by the datasheet.
-    */
-    i2c->STAT0 &= ~ERROR_FLAGS;
-    _stop(i2c);
-    return -ETIMEDOUT;
+
+    /* if I2C_NOSTOP flag is not set, send STOP condition */
+    if (!(flags & I2C_NOSTOP)) {
+        /* send STOP condition */
+        res = _i2c_stop_cmd(dev);
+        if (res != 0) {
+            return res;
+        }
+    }
+
+    return _wait_for_bus(dev);
 }
 
-static int _stop(I2C_Type *i2c)
+void _i2c_transfer_timeout(void *arg)
 {
-    /* send STOP condition */
-    DEBUG("[i2c] stop: Generate stop condition\n");
-    i2c->CTL0 &= ~(I2C_CTL0_ACKEN_Msk);
-    i2c->CTL0 |= I2C_CTL0_STOP_Msk;
-    uint16_t tick = TICK_TIMEOUT;
-    while ((i2c->CTL0 & I2C_CTL0_STOP_Msk) && tick--) {}
-    if (!tick) {
-        return -ETIMEDOUT;
-    }
-    DEBUG("[i2c] stop: Stop condition succeeded\n");
-    if (_wait_for_bus(i2c) < 0) {
-        return -ETIMEDOUT;
+    i2c_t dev = (i2c_t)(uintptr_t)arg;
+
+    /* set result to timeout */
+    _i2c_dev[dev].state = I2C_TIMEOUT;
+
+    /* wake up the thread that is waiting for the results */
+    mutex_unlock(&_i2c_dev[dev].irq_lock);
+}
+
+static void _wait_for_irq(i2c_t dev)
+{
+#if defined(MODULE_ZTIMER_USEC)
+    ztimer_t timer = { .callback = _i2c_transfer_timeout,
+                       .arg = (void *)dev };
+    uint32_t timeout = ((I2C_TIMEOUT_CYCLES * MHZ(1)) / i2c_config[dev].speed) + 1;
+    ztimer_set(ZTIMER_USEC, &timer, timeout);
+#elif defined(MODULE_ZTIMER_MSEC)
+    ztimer_t timer = { .callback = _i2c_transfer_timeout,
+                       .arg = (void *)dev };
+    uint32_t timeout = ((I2C_TIMEOUT_CYCLES * KHZ(1)) / i2c_config[dev].speed) + 1;
+    ztimer_set(ZTIMER_MSEC, &timer, timeout);
+#else
+#warning "I2C timeout handling requires to use ztimer_msec or ztimer_usec"
+#endif
+
+    /* enable only event and error interrupts, buffer interrupts are only used in read */
+    i2c_config[dev].dev->CTL1 |= I2C_INT_EV_ERR_FLAGS;
+
+    /* wait for buffer, event or error interrupt */
+    mutex_lock(&_i2c_dev[dev].irq_lock);
+
+#if defined(MODULE_ZTIMER_USEC)
+    ztimer_remove(ZTIMER_USEC, &timer);
+#elif defined(MODULE_ZTIMER_MSEC)
+    ztimer_remove(ZTIMER_MSEC, &timer);
+#endif
+
+    if (_i2c_dev[dev].state == I2C_TIMEOUT) {
+        DEBUG("error: timeout, dev=%d\n", dev);
     }
-    DEBUG("[i2c] stop: Bus is free\n");
-    return 0;
 }
 
-static inline int _wait_for_bus(I2C_Type *i2c)
+#define TICK_TIMEOUT    0x00000fffUL
+
+static inline int _wait_for_bus(i2c_t dev)
 {
     uint16_t tick = TICK_TIMEOUT;
-    while ((i2c->STAT1 & I2C_STAT1_I2CBSY_Msk) && tick--) {}
+    /* short busy waiting for the bus becoming free (I2CBSY is cleared) */
+    while ((i2c_config[dev].dev->STAT1 & I2C_STAT1_I2CBSY_Msk) && tick--) {}
```

Comment:
So that's just the delay to power on the I2C peripheral, right?

Reply:
No, it is only a very short delay from setting the STOP bit to sending the STOP condition on the bus. `I2C_STAT1_I2CBSY` is cleared by the controller as soon as the STOP condition is sent. Unfortunately this does not trigger an interrupt, so polling is the only way.